[OpenMP] Add support for lowering Critical Construct

kiranchandramohan opened the pull request:
The patch adds lowering for the Critical Construct.

The patch uses some code from @Sameeranjoshi's patch https://reviews.llvm.org/D93051.

jeanPerier reviewed (commented):
> Is that OK? Also what should be the mangling if any?

My concern would be potential conflicts with other program symbols after remangling (for instance, a `BIND(C, name="help")` variable somewhere else). Even if it is illegal as you are pointing out, I am not sure the front-end enforces it, so adding a bit more of protection seems safer to me and would also help navigating through the globals/symbols in  the object files.

Is clang also creating a global of a certain name for the C/C++ equivalent program if there is any ?

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -364,9 +364,12 @@ getScheduleModifiers(const Fortran::parser::OmpScheduleClause &x) {
+          modType2->v.v !=
+              Fortran::parser::OmpScheduleModifierType::ModType::Simd)
         return translateModifier(modType2->v);
+      else
```
It looks like `else` should be deleted.

kiranchandramohan replied:
Assuming this is a coding style comment. Done.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -600,7 +603,48 @@ void Fortran::lower::genOpenMPConstruct(
-            TODO(converter.getCurrentLocation(), "OpenMPCriticalConstruct");
+            auto &firOpBuilder = converter.getFirOpBuilder();
+            auto currentLocation = converter.getCurrentLocation();
+            mlir::omp::CriticalOp criticalOp;
```
You could even use a lambda to initialize it directly (if needed), as in

```c++
auto criticalOp = [&]() -> mlir::omp::CriticalOp { ... }();
```

kiranchandramohan replied:
Thanks, I have used @schweitzpgi's suggestion.

schweitzpgi commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -204,6 +204,43 @@ convertOmpMaster(Operation &opInst, llvm::IRBuilderBase &builder,
+    hint =
+        llvm::ConstantInt::get(llvm::Type::getInt32Ty(llvmContext),
+                               static_cast<int>(criticalOp.hint().getValue()));
+  }
```
Yes. The assumption is that upstreamed changes are more correct than whatever was cherry-picked in conflicts other than lowering and optimization.

kiranchandramohan replied:
:) We need John the merciful.

I think we have now added a few changes to the mlir repo, I will upstream all changes possible.

schweitzpgi reviewed (commented):
On the mangling question: this can use a simple bare name, the unidirectional mangling (`_QQ...`), just "pretend" these are global constants, or we could add a different bijective mangling for this case. Personally, I'd lean towards the last in order to avoid any unpleasant compiler crashes due to a possible IR symbol binding collision.

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -600,7 +603,48 @@ void Fortran::lower::genOpenMPConstruct(
+                const auto hintValue = Fortran::evaluate::ToInt64(*expr);
+                hint = mlir::omp::SyncHintKindAttr::get(
+                    firOpBuilder.getContext(),
+                    mlir::omp::symbolizeSyncHintKind(*hintValue).getValue());
```
Good point. Yes, we can break. There will only be one hint clause and the standard prescribes this through the syntax.
```
!$omp critical [(name) [[,] hint(hint-expression)] ]
```
I have added a change to the OMP.td file to enforce this restriction.

kiranchandramohan commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1454,6 +1464,12 @@ void OmpAttributeVisitor::ResolveOmpName(
+  } else if (ompFlag == Symbol::Flag::OmpCriticalLock) {
+    // Create a new symbol.
+    const auto pair{GetContext().scope.try_emplace(
+        name.source, Attrs{}, UnknownDetails{})};
```
That patch has been lying idle for a while since @Sameeranjoshi moved to another team. But I hear that he is back. So there is a good chance that this will land soon with some changes (including this change). Otherwise, I will take the patch forward.

kiranchandramohan commented:
> My concern would be potential conflicts with other program symbols after remangling (for instance, a BIND(C, name="help") variable somewhere else). Even if it is illegal as you are pointing out, I am not sure the front-end enforces it, so adding a bit more of protection seems safer to me and would also help navigating through the globals/symbols in the object files.

You are suggesting a mangling for openmp critical here, right?

> Is clang also creating a global of a certain name for the C/C++ equivalent program if there is any ?

Yes. For e.g. Clang creates the following for a critical section with name "nm1".
@.gomp_critical_user_nm1.var = common global [8 x i32] zeroinitializer

> On the mangling question: this can use a simple bare name, the unidirectional mangling (_QQ...), just "pretend" these are global constants, or we could add a different bijective mangling for this case. Personally, I'd lean towards the last in order to avoid any unpleasant compiler crashes due to a possible IR symbol binding collision.

OK. I will try to add a different mangling.

I have another question/point for discussion here. We currently create the following global in FIR for a critical section with name (for eg help).
```
fir.global @help constant : !fir.char<1,4>
```
This gets converted into the following in LLVM IR.
```
@help = external constant [4 x i8]
```
The OpenMP IRBuilder also creates the following for use in the runtime call below. Now `@help = external constant [4 x i8]` in LLVM IR created from the fir.global is probably extraneous and irrelevant. But at the same time, the fir.global is necessary to enforce the naming issue in FIR. 
```
@.gomp_critical_user_.var = common global [8 x i32] zeroinitializer
  call void @__kmpc_critical_with_hint(%struct.ident_t* @1, i32 %omp_global_thread_num, [8 x i32]* @.gomp_critical_user_help.var, i32 2), !dbg !10
  ```
  
  Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

kiranchandramohan commented:
ping.

jeanPerier commented:
> You are suggesting a mangling for openmp critical here, right?

Yes, I think it is safer.

> Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

Do you mean to somehow reserve the name without creating any fir.global ? If so, I do not think it is possible because the IR itself is used for the bookeeping of fir.global. Couldn't the OpenMP IRBuilder somehow recognized the global created for the critical and replace it or delete it (or maybe the mlir::omp::CriticalOp rewrite in MLIR )?

schweitzpgi commented:
> Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?

I'm not entirely sure what you're looking for. It is possible to create globals with different linkages and/or without initializers. There are some restrictions though from what the LLVM-IR dialect of MLIR will support or not, so maybe that is what you're experiencing?

kiranchandramohan commented:
> > Is it possible to just declare a name in FIR to represent a global (like a function declaration) but not create a variable (because this will ultimately get created in LLVM IR)?
> 
> I'm not entirely sure what you're looking for. It is possible to create globals with different linkages and/or without initializers. There are some restrictions though from what the LLVM-IR dialect of MLIR will support or not, so maybe that is what you're experiencing?

Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.

schweitzpgi commented:
> Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
> This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.

Thanks for the explanation, @kiranchandramohan. Makes sense. I assume the OpenMP IRBuilder naming scheme won't conflict with ad hoc Fortran naming schemes that might be implemented for interoperating with other Fortran compilers, etc.

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -600,7 +602,51 @@ void Fortran::lower::genOpenMPConstruct(
           [&](const Fortran::parser::OpenMPCriticalConstruct
                   &criticalConstruct) {
-            TODO(converter.getCurrentLocation(), "OpenMPCriticalConstruct");
+            auto &firOpBuilder = converter.getFirOpBuilder();
```
I have made this change as the Critical function lowering has become a bit bigger.

Please note that this is only suggestive and we usually stick to mlir/llvm coding guidelines in the Lower and Optimize directories.

PeixinQiao replied:
Thanks for the explanation.

kiranchandramohan commented on flang/test/Lower/OpenMP/critical.f90:
```diff
@@ -0,0 +1,42 @@
+!LLVMIRDialect: omp.terminator
+!LLVMIR: call void @__kmpc_end_critical({{.*}}, {{.*}}, {{.*}} @{{.*}}help.var)
+!$OMP END CRITICAL(help)
+
```
Done.

kiranchandramohan commented on mlir/test/Target/LLVMIR/openmp-llvm.mlir:
```diff
@@ -489,4 +489,28 @@ llvm.func @test_omp_wsloop_dynamic_monotonic(%lb : i64, %ub : i64, %step : i64)
+    omp.terminator
+  }
+  // CHECK: call void @__kmpc_end_critical({{.*}}critical_user_.var{{.*}})
+
```
Done.

kiranchandramohan commented:
> > Thanks @schweitzpgi for your question. Ultimately I decided to handle this in the OpenMP dialect itself. In the MLIR layer, we need names of OpenMP Critical constructs to be unique. This is handled by creating a "declare critical" operation that will declare names of critical sections. The existence of such an operation also removes the need for mangling. The OpenMP IRBuilder will create a variable (with a name derived from the critical section name) that functions as a lock for the critical section. We only need to pass the name of the critical section to the OpenMP IRBuilder.
> > This is approved upstream (https://reviews.llvm.org/D108713). I will cherry-pick it here after landing.
> 
> Thanks for the explanation, @kiranchandramohan. Makes sense. I assume the OpenMP IRBuilder naming scheme won't conflict with ad hoc Fortran naming schemes that might be implemented for interoperating with other Fortran compilers, etc.

Unlikely to conflict with Fortran names since there will be a `gomp` and `critical` prefix, e.g. for a critical section with name `help` the variable will have the name `@.gomp_critical_user_help.var`.

I have now cherry-picked https://reviews.llvm.org/D108713 (and quite a few dependent commits to make the build clean).

Have also addressed all review comments.

kiranchandramohan commented:
ping. Can this be approved if OK?

kiranchandramohan commented:
The first 15 commits in the PR are cherry-picked from upstream llvm-project. Would it be better to have these cherry-picked commits in a separate PR for ease of review and approval?

Also, in general can we have a simplified process for approving and merging cherry-picked commits from upstream llvm-project?

jeanPerier commented:
#1055 has been merged, can this one be rebased ?

kiranchandramohan commented:
Thanks for the ping @jeanPerier. Rebased.

jeanPerier reviewed (approved):
Thanks for rebasing @kiranchandramohan .
Lowering logs good to me. Please make sure to upstream the parser/semantics changes, it looks like they are not yet in LLVM.

kiranchandramohan commented:
Sure @jeanPerier. AMD folks are working on the parser/semantics part in upstream llvm. I will ask them to cherry-pick it here and solve any conflicts.

@schweitzpgi can this be merged?
